dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/validation/validator/event/SecurityTrackerEventValidator.java

```diff
@@ -62,141 +62,69 @@
+      boolean isCreatableInSearchScope =
           strategy.isCreate()
               ? event.isCreatableInSearchScope()
-              : preheatEvent.isCreatableInSearchScope(),
-          bundle.getUser());
-    }
+              : preheatEvent.isCreatableInSearchScope();
 
-    UID teUid = getTeUidFromEvent(bundle, event, program);
+      checkEventOrgUnitWriteAccess(
+          reporter, event, organisationUnit, isCreatableInSearchScope, bundle.getUser());
+      checkProgramStageWriteAccess(reporter, event, programStage, bundle.getUser());
+      checkProgramReadAccess(reporter, event, programStage.getProgram(), bundle.getUser());
+      checkTeTypeReadAccess(reporter, event, programStage.getProgram(), bundle.getUser());
+      checkOwnership(
+          reporter, event, teUid, ownerOrgUnit, programStage.getProgram(), bundle.getUser());
+      checkWriteCategoryOptionComboAccess(reporter, event, categoryOptionCombo, bundle.getUser());
 
-    CategoryOptionCombo categoryOptionCombo =
-        bundle.getPreheat().getCategoryOptionCombo(event.getAttributeOptionCombo());
-    OrganisationUnit ownerOrgUnit = getOwnerOrganisationUnit(preheat, teUid, program);
-    // Check acting user is allowed to change existing/write event
-    if (strategy.isUpdateOrDelete()) {
-      TrackedEntity trackedEntity = preheatEvent.getEnrollment().getTrackedEntity();
-      validateUpdateAndDeleteEvent(
-          reporter,
-          bundle,
-          event,
-          preheatEvent,
-          trackedEntity == null ? null : UID.of(trackedEntity),
-          ownerOrgUnit,
-          bundle.getUser());
-    } else {
-      validateCreateEvent(
-          reporter,
-          bundle,
-          event,
-          categoryOptionCombo,
-          programStage,
-          teUid,
-          organisationUnit,
-          ownerOrgUnit,
-          program,
-          event.isCreatableInSearchScope());
+      if (strategy.isUpdate()) {
+        checkCompletablePermission(reporter, event, preheatEvent, bundle.getUser());
+      }
     }
   }
 
-  private void validateCreateEvent(
+  private void checkCompletablePermission(
       Reporter reporter,
-      TrackerBundle bundle,
-      org.hisp.dhis.tracker.imports.domain.Event event,
-      CategoryOptionCombo categoryOptionCombo,
-      ProgramStage programStage,
-      UID teUid,
-      OrganisationUnit organisationUnit,
-      OrganisationUnit ownerOrgUnit,
-      Program program,
-      boolean isCreatableInSearchScope) {
-    boolean noProgramStageAndProgramIsWithoutReg =
-        programStage == null && program.isWithoutRegistration();
-
-    programStage =
-        noProgramStageAndProgramIsWithoutReg ? program.getProgramStageByStage(1) : programStage;
-
-    checkEventWriteAccess(
-        reporter,
-        bundle,
-        event,
-        programStage,
-        organisationUnit,
-        ownerOrgUnit,
-        categoryOptionCombo,
-        // TODO: Calculate correct `isCreateableInSearchScope` value
```

Comment:
Is this todo fixed or was it not relevant anymore?

Reply:
Not relevant anymore, it was fixed at some point, as now the variable is correctly calculated, but the TODO was not removed.